subsys/power/device_pm.c

```diff
@@ -0,0 +1,238 @@
+		dev->pm->status = DEVICE_PM_REQ_STAT_ERROR;
+	} else {
+		dev->pm->status = DEVICE_PM_REQ_STAT_DONE;
+	}
+out:
+	/* Notify the driver through alert */
+	if (dev->pm->alert.handler) {
+		k_alert_send(&dev->pm->alert);
+	}
+	k_sem_give(&dev->pm->lock);
+}
+
+/*
+ * Set the device PM state to target PM state based on
+ * the current state. If the PM request is of Async type
+ * then a work item will be submitted to PM worker queue
+ * else it will call device_set_power_state() to set the
+ * device PM state to target state.
+ */
+static int device_pm_set_state(struct device *dev,
+			       u32_t target_state, u32_t pm_flags)
+{
+	int ret;
+	u32_t current_state;
+
+	/* Get current device power state */
+	ret = device_get_power_state(dev, &current_state);
+	if (ret) {
+		LOG_ERR("Get power state error(%d)\n", ret);
+		goto out;
+	}
+
+	/* Return if current state is same as target state */
+	if (current_state == target_state) {
+		goto out;
+	}
+
+	/* Cancel already queued work */
+	if (dev->pm->status == DEVICE_PM_REQ_STAT_INIT) {
+		k_delayed_work_cancel(&dev->pm->dwork);
+		dev->pm->status = DEVICE_PM_REQ_STAT_CANCEL;
+		/* Notify the driver through alert */
+		if (dev->pm->alert.handler) {
+			k_alert_send(&dev->pm->alert);
+		}
+	}
+
+	/* Schedule the worker incase of async request */
+	if (pm_flags & DEVICE_PM_ASYNC) {
+		dev->pm->request = target_state;
+		dev->pm->status = DEVICE_PM_REQ_STAT_INIT;
+		k_delayed_work_submit_to_queue(&devpm_workq,
+					&dev->pm->dwork, dev->pm->timeout);
+	} else {
+		ret = device_set_power_state(dev, target_state);
+		if (ret) {
+			LOG_ERR("Set power state(%d) error(%d)\n",
+							target_state, ret);
+		}
+	}
+
+out:
+	return ret;
+}
+
+int device_pm_get(struct device *dev)
+{
+	int ret;
+
+	if (atomic_inc(&dev->pm->usage) >= UINT_MAX) {
+		LOG_ERR("usage count overflow error!!\n");
+		return -ERANGE;
+	}
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (!dev->pm->enable) {
+		k_sem_give(&dev->pm->lock);
+		return 0;
+	}
+
+	ret = device_pm_set_state(dev,
+			DEVICE_PM_ACTIVE_STATE, DEVICE_PM_ASYNC);
+	k_sem_give(&dev->pm->lock);
+	return ret;
+}
+
+int device_pm_get_sync(struct device *dev)
+{
+	int ret;
+
+	if (atomic_inc(&dev->pm->usage) >= UINT_MAX) {
+		LOG_ERR("usage count overflow error!!\n");
+		return -ERANGE;
+	}
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (!dev->pm->enable) {
+		k_sem_give(&dev->pm->lock);
+		return 0;
+	}
+
+	ret = device_pm_set_state(dev, DEVICE_PM_ACTIVE_STATE, 0);
+	k_sem_give(&dev->pm->lock);
+	return ret;
+}
+
+int device_pm_put(struct device *dev)
+{
+	int ret;
+
+	if (atomic_dec(&dev->pm->usage) <= 0) {
+		LOG_ERR("usage count underflow error!!\n");
+		return -ERANGE;
+	}
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (!dev->pm->enable) {
+		k_sem_give(&dev->pm->lock);
+		return 0;
+	}
+
+	ret = device_pm_set_state(dev,
+			DEVICE_PM_SUSPEND_STATE, DEVICE_PM_ASYNC);
+	k_sem_give(&dev->pm->lock);
+	return ret;
+}
+
+int device_pm_put_sync(struct device *dev)
+{
```

Comment:
Both synchronous and asynchronous variants of `pm_put` (`pm_get`) are identical except for the flags passed to the underlying `set_state`.  Put the put (get)  implementation in a static function that takes the flags as a parameter and call it from the sync and async variants with the appropriate argument, so we don't have to worry about the locking logic being right in one case but not the other.  You might even be able to pass the state as an argument and combine all four into one, though that might get confusing.

Reply:
Ok thanks.

Reply:
this is fixed.